### DISCUSSION TITLE
React 18 breaking change

### DISCUSSION THREAD

**Discussion by frattaro** [vercel/next.js#36533] [Help]
Repro here: https://github.com/frattaro/nextjs-react18-issue

This is the interesting part (_app.js):

```
import {useState} from "react";


function MyApp({ Component, ssrParam, pageProps }) {
  const [test] = useState(ssrParam || window.clientInformation);
  return <Component {...pageProps} />
}

export default MyApp
```

Works with react 17.0.2, fails with react 18.1.0

Does anyone know why?

The repro is trivial, but the real use case is around nonces and emotion cache and it'd make the issue really cloudy if I had included it.

Thanks :)

**Comment by icyJoseph** [comment]
Really strange issue, of course, we ought to avoid direct window access, but even if we put a guard for window (`typeof window !== 'undefined'`), there's clearly some type of behaviour change here between 17 and 18.

I added a log inside `_app`, `console.log({ ssrParam }, "window:", typeof window)`, and a `console.count` for `enhanceApp`:

```shell
event - compiled client and server successfully in 55 ms (144 modules)
{ ssrParam: undefined } window: undefined
EnhanceApp: 1
{ ssrParam: 'ssrParam' } window: undefined
```

To no surprise the server renders a frame where both the `ssrParam` and `window` are `undefined`, and only then the `enhancedApp` is called. This ought to be something with the way the server html is generated nowadays, as a stream rather just flushed synchronously.

The client logs once `{ssrParam: undefined} 'window:' 'object'`.

All of this with Strict Mode `true`.

## Edit with React 18.0.0

```shell
{ ssrParam: undefined } window: undefined
EnhanceApp: 1
{ ssrParam: 'ssrParam' } window: undefined
{ ssrParam: undefined } window: undefined
EnhanceApp: 2
{ ssrParam: 'ssrParam' } window: undefined
```

There were some fixes for hydration on React 18.1.0, which make this even more mysterious 😅

**Comment by frattaro** [comment]
As a note, turning off strict mode does not fix this. (I saw this suggestion floating around twitter)

**Comment by hamlim** [comment]
I've seen in a few places around hydration issues on the React repo that the initial client render should exactly match the result of the SSR render pass, so for example that `useState` initial value sometimes using a param and sometimes using a value from window would possibly be unsafe if the param **isn't** provided as the same value in both cases.
